src/test/java/com/pusher/client/channel/impl/ChannelImplTest.java

```diff
@@ -201,6 +203,36 @@ public void testUnbindWhenInUnsubscribedStateThrowsException() {
         channel.updateState(ChannelState.UNSUBSCRIBED);
         channel.unbind(EVENT_NAME, mockListener);
     }
+    @Test
+    public void testMetaDataIsExtractedFromMessageAndPassedToSingleListener() {
+        final ArgumentCaptor<PusherEvent> argument = ArgumentCaptor.forClass(PusherEvent.class);
+
+        final String eventName = "client-my-event";
+        ChannelEventListener mockListener = getEventListener();
+
+        channel = newInstance(getChannelName());
+        channel.bind(eventName, mockListener);
+        channel.onMessage(eventName, "{\"event\":\"client-my-event\",\"data\":\"{\\\"fish\\\":\\\"chips\\\"}\",\"metadata-key\":\"42\"}");
+
+        verify(mockListener).onEvent(eq(getChannelName()), eq(eventName), eq("{\"fish\":\"chips\"}"), argument.capture());
+        assertEquals("42", argument.getValue().getProperty("metadata-key"));
+    }
+
+    @Test
+    public void testUserIdIsExtractedFromMessageAndPassedToSingleListenerWithHelperMethod() {
+        final ArgumentCaptor<PusherEvent> argument = ArgumentCaptor.forClass(PusherEvent.class);
+
+        final String eventName = "client-my-event";
+        ChannelEventListener mockListener = getEventListener();
+
+        channel = newInstance(getChannelName());
+        channel.bind(eventName, mockListener);
+        channel.onMessage(eventName, "{\"event\":\"client-my-event\",\"data\":\"{\\\"fish\\\":\\\"chips\\\"}\",\"user_id\":\"Kuzya\"}");
+
+        verify(mockListener).onEvent(eq(getChannelName()), eq(eventName), eq("{\"fish\":\"chips\"}"), argument.capture());
+        assertEquals("Kuzya", argument.getValue().getUserId());
+    }
+
```

Comment:
It might be worth adding some other tests to check that the `PusherEvent` contains the correct channel name, event and data properties.